File: bun-version/src/providers/zai.ts
import { BaseSearchProvider } from "./base";
import type { WebSearchResult } from "../types";

export class ZaiProvider extends BaseSearchProvider {
  readonly name = "zai";
  readonly toolName = "zai-search";
  readonly description = "智谱 AI 联网搜索";
  readonly requiredParams = ["zai-apiKey"];
  readonly optionalParams = [
    {
      name: "search_engine",
      description: "搜索引擎 (search_std/search_pro/search_pro_sogou/search_pro_quark)",
      defaultValue: "search_std",
    },
    { name: "count", description: "返回结果数 (1-50)", defaultValue: "10" },
    {
      name: "search_recency_filter",
      description: "时间过滤 (oneDay/oneWeek/oneMonth/oneYear/noLimit)",
      defaultValue: "noLimit",
    },
    {
      name: "content_size",
      description: "内容长度 (medium/high)",
      defaultValue: "medium",
    },
  ];

  async search(
    query: string,
    urlParams: URLSearchParams
  ): Promise<WebSearchResult> {
    const apiKey = urlParams.get("zai-apiKey")!;
    const searchEngine = this.getParam(
      urlParams,
      "search_engine",
      "search_std"
    )!;
    const count = parseInt(this.getParam(urlParams, "count", "10")!);
    const recencyFilter = this.getParam(
      urlParams,
      "search_recency_filter",
      "noLimit"
    )!;
    const contentSize = this.getParam(urlParams, "content_size", "medium");

    console.log(`[zai-search] query: "${query}", engine: ${searchEngine}`);

    const resp = await fetch(
      "https://open.bigmodel.cn/api/paas/v4/web_search",
      {
        method: "POST",
        headers: {
          Authorization: `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          search_query: query,
          search_engine: searchEngine,
          search_intent: false,
          count,
          search_recency_filter: recencyFilter,
          ...(contentSize ? { content_size: contentSize } : {}),
        }),
      }
    );

    if (!resp.ok) {
      throw new Error(
        `zai-search API error: ${resp.status} ${await resp.text()}`
      );
    }

    const data = (await resp.json()) as any;
    const results = data.search_result || [];
    const text = results
      .map(
        (r: any, i: number) =>
          `[${i + 1}] ${r.title}\n${r.content}\n来源: ${r.link}`
      )
      .join("\n\n");

    return {
      content: [{ type: "text", text: text || "未找到搜索结果" }],
    };
  }
}

File: src/provider/zai.rs
use super::{SearchProvider, WebSearchResult};
use serde_json::json;
use std::collections::HashMap;

pub struct ZaiProvider;

#[async_trait::async_trait]
impl SearchProvider for ZaiProvider {
    fn name(&self) -> &str {
        "zai"
    }

    fn tool_name(&self) -> &str {
        "zai-search"
    }

    fn description(&self) -> &str {
        "智谱 AI 联网搜索"
    }

    fn required_params(&self) -> &[&str] {
        &["zai-apiKey"]
    }

    async fn search(
        &self,
        query: &str,
        params: &HashMap<String, String>,
        client: &reqwest::Client,
    ) -> Result<WebSearchResult, String> {
        let api_key = params
            .get("zai-apiKey")
            .ok_or_else(|| "Missing zai-apiKey".to_string())?;

        let search_engine = params
            .get("zai-search_engine")
            .map(|s| s.as_str())
            .unwrap_or("search_std");

        let count = params
            .get("zai-count")
            .and_then(|s| s.parse::<u32>().ok())
            .unwrap_or(10);

        let recency_filter = params
            .get("zai-search_recency_filter")
            .map(|s| s.as_str())
            .unwrap_or("noLimit");

        let content_size = params
            .get("zai-content_size")
            .map(|s| s.as_str())
            .unwrap_or("medium");

        println!(
            "[zai-search] query: \"{}\", engine: {}",
            query, search_engine
        );

        let mut body = json!({
            "search_query": query,
            "search_engine": search_engine,
            "search_intent": false,
            "count": count,
            "search_recency_filter": recency_filter,
        });

        if !content_size.is_empty() {
            if let Some(obj) = body.as_object_mut() {
                obj.insert(
                    "content_size".to_string(),
                    serde_json::Value::String(content_size.to_string()),
                );
            }
        }

        let res = client
            .post("https://open.bigmodel.cn/api/paas/v4/web_search")
            .bearer_auth(api_key)
            .json(&body)
            .send()
            .await
            .map_err(|e| format!("Network error: {}", e))?;

        if !res.status().is_success() {
            let err_text = res.text().await.unwrap_or_default();
            return Err(format!("zai-search API error: {}", err_text));
        }

        let data: serde_json::Value = res
            .json()
            .await
            .map_err(|e| format!("Failed to parse response: {}", e))?;

        let results = data["search_result"].as_array();
        let formatted_text = match results {
            Some(items) if !items.is_empty() => items
                .iter()
                .enumerate()
                .map(|(i, r)| {
                    let title = r["title"].as_str().unwrap_or("");
                    let content = r["content"].as_str().unwrap_or("");
                    let link = r["link"].as_str().unwrap_or("");
                    format!("[{}] {}\n{}\n来源: {}", i + 1, title, content, link)
                })
                .collect::<Vec<_>>()
                .join("\n\n"),
            _ => "未找到搜索结果".to_string(),
        };

        Ok(WebSearchResult::text(formatted_text))
    }
}
